In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [2]:
calls = pd.read_excel('Calls (Final).xlsx', dtype={'Id': str, "CONTACTID": str})
spend = pd.read_excel('Spend (Final).xlsx')
deals = pd.read_excel('Deals (Final).xlsx', dtype={'Id': str, 'Contact Name': str})

In [3]:
deals.head()

Id Deal Owner Name Closing Date            Quality  \
0  5805028000056864695        Ben Hall          NaT  E - Non Qualified   
1  5805028000056859489   Ulysses Adams          NaT  E - Non Qualified   
2  5805028000056832357   Ulysses Adams   2024-06-21     D - Non Target   
3  5805028000056824246        Eva Kent   2024-06-21  E - Non Qualified   
4  5805028000056873292        Ben Hall   2024-06-21     D - Non Target   

      Stage     Lost Reason                  Campaign          Source  \
0  New Lead        Not lost             03.07.23women    Facebook Ads   
1  New Lead        Not lost                       NaN         Organic   
2      Lost      Non target                engwien_AT  Telegram posts   
3      Lost  Invalid number  04.07.23recentlymoved_DE    Facebook Ads   
4      Lost      Non target              discovery_DE      Google Ads   

          Payment Type        Product Education Type        Created Time  \
0          One Payment  Not specified        Unknown 2024-06-21 15:30:00   
1  No Initial Payments  Web Developer        Morning 2024-06-21 15:23:00   
2          One Payment  Not specified        Unknown 2024-06-21 14:45:00   
3          One Payment  Not specified        Unknown 2024-06-21 13:32:00   
4          One Payment  Not specified        Unknown 2024-06-21 13:21:00   

   Course duration  Months of study  Initial Amount Paid  Offer Total Amount  \
0                0              NaN                  NaN                 NaN   
1                6              NaN                  0.0              2000.0   
2                0              NaN                  NaN                 NaN   
3                0              NaN                  NaN                 NaN   
4                0              NaN                  NaN                 NaN   

          Contact Name         City Level of Deutsch  SLA in Seconds  
0  5805028000056849495  Not defined      Not defined               0  
1  5805028000056834471  Not defined      Not defined               0  
2  5805028000056854421  Not defined      Not defined            1603  
3  5805028000056889351  Not defined      Not defined            3604  
4  5805028000056876176  Not defined      Not defined            3192

In [4]:
spend.head()

Date        Source               Campaign  Impressions  Spend  Clicks  \
0 2023-07-03    Google Ads         gen_analyst_DE            6   0.00       0   
1 2023-07-03    Google Ads  performancemax_eng_DE            4   0.01       1   
2 2023-07-03  Facebook Ads          03.07.23women          187   3.30       6   
3 2023-07-03  Facebook Ads          03.07.23women            4   0.02       1   
4 2023-07-03  Facebook Ads        02.07.23wide_DE           61   0.58       0   

   AdGroup  
0  Unknown  
1  Unknown  
2    women  
3    women  
4     wide

In [5]:
# Set display format for floats
pd.set_option('display.float_format', '{:,.2f}'.format)

# Filter out 'Not specified' products
filtered_deals = deals[deals['Product'] != 'Not specified']

# Leads (UA): Count of unique leads
leads = filtered_deals.groupby(['Product', 'Education Type'], observed=True)['Contact Name'].nunique().rename('Leads')

# Customers (Buyers): Count of unique customers who completed payment
customers = filtered_deals[filtered_deals['Stage'] == 'Payment Done'] \
    .groupby(['Product', 'Education Type'], observed=True)['Contact Name'].nunique().rename('Customers')

# Successful Deals: Total number of 'Payment Done' deals
successful_deals = filtered_deals[filtered_deals['Stage'] == 'Payment Done'] \
    .groupby(['Product', 'Education Type'], observed=True)['Id'].count().rename('Successful Deals')

# Gross Revenue: Sum of 'Offer Total Amount' for successful deals
gross_revenue = filtered_deals[filtered_deals['Stage'] == 'Payment Done'] \
    .groupby(['Product', 'Education Type'], observed=True)['Offer Total Amount'].sum().rename('Revenue')

# Merge all metrics into one DataFrame
metrics = leads.to_frame().join([customers, successful_deals, gross_revenue], how='left').reset_index()

# Total marketing spend
total_spend = spend['Spend'].sum()

# Add calculated unit economics metrics
metrics['Conversion Rate'] = metrics['Customers'] / metrics['Leads']
metrics['Total Marketing Spend'] = total_spend
metrics['CPA'] = metrics['Total Marketing Spend'] / metrics['Leads']           # Cost Per Acquisition (lead)
metrics['CAC'] = metrics['Total Marketing Spend'] / metrics['Customers']       # Customer Acquisition Cost
metrics['AOV'] = metrics['Revenue'] / metrics['Customers']                     # Average Order Value
metrics['APC'] = metrics['Successful Deals'] / metrics['Customers']            # Average Purchases per Customer
metrics['CLTV'] = metrics['AOV'] * metrics['APC']                               # Customer Lifetime Value
metrics['LTV'] = metrics['Revenue'] / metrics['Leads']                         # Revenue per Lead
metrics['Contribution Margin'] = (metrics['CLTV'] - metrics['CAC']) * metrics['Customers']

# Sort by number of customers descending
metrics_sorted = metrics.sort_values(by='Customers', ascending=False)

# Reorder columns
metrics_sorted = metrics_sorted[[
    'Product', 'Education Type', 'Leads', 'Customers', 'Conversion Rate',
    'Total Marketing Spend', 'CPA', 'CAC', 'Revenue', 'AOV', 'APC',
    'CLTV', 'LTV', 'Contribution Margin', 'Successful Deals'
]]

# Format large numbers
def format_number(value):
    return f'{value:,.0f}'.replace(',', ' ')

# Create styled table
styled_metrics = metrics_sorted.style.format({
    'Leads': format_number,
    'Customers': format_number,
    'Revenue': format_number,
    'Total Marketing Spend': format_number,
    'CPA': format_number,
    'CAC': format_number,
    'AOV': format_number,
    'CLTV': format_number,
    'LTV': format_number,
    'Contribution Margin': format_number,
    'Conversion Rate': '{:.2%}'.format,
    'APC': '{:,.2f}'.format,
    'Successful Deals': format_number
}).hide(axis='index')

# Display the final styled table
display(styled_metrics)


🔍 Key Insights

1. Top Performers in Conversion and Profitability

Digital Marketing (Morning)

Conversion Rate: 92.73%

Revenue: 17.1M

LTV: 10,036

Contribution Margin: 17.2M (highest overall)

CPA: 88 — very efficient
✅ This is the strongest product-format combination. Excellent conversion, low acquisition costs, and top-line contribution make it a core growth driver.

UX/UI Design (Morning)

Conversion Rate: 97.78% (highest overall)

LTV: 10,332

CAC: 189

Contribution Margin: 8.3M
✅ Exceptionally high conversion and retention. Strong economics — deserves increased marketing support and lead volume.

Web Developer (Morning)

Conversion Rate: 92.02%

LTV: 5,089

Contribution Margin: 2.76M
✅ Smaller in scale but shows strong fundamentals. AOV and LTV are solid. Good candidate for scaling.

2. Underperformers and Risk Areas

UX/UI Design (Evening)

Conversion Rate: 63.00%

CAC: 1,187 (highest)

Contribution Margin: 347K
⚠️ Evening version is highly inefficient compared to morning. Significantly worse CAC and LTV.

Digital Marketing (Evening)

LTV: 3,181 vs. 10,036 (morning)

CAC: 709 vs. 95 (morning)
⚠️ Evening format dilutes profitability. High costs with comparatively low returns.

Low-Volume, High-Cost Courses (Find Yourself in IT, Web Developer Evening, Data Analytics)
🚫 These courses have extremely high CAC (up to 149K), minimal revenue, and negative margins.

📈 Opportunities for Growth

1. Scale High-Performing Morning Programs
Focus on Digital Marketing (Morning) and UX/UI Design (Morning). Low CAC, high conversion, and excellent margins. Expand marketing, lead generation, and course capacity.
Expected Impact: Higher revenue and profitability through proven performers.

2. Optimize or Phase Out Evening Offerings
Test improvements to UX/UI Design (Evening) — curriculum, pricing, or instructors. Set KPIs for evaluation.
Expected Impact: Either turn around performance or reallocate resources to better use.

3. Grow Web Developer (Morning) Strategically
Strong unit economics, but underutilized. Launch targeted lead generation campaigns.
Expected Impact: Mid-level revenue growth with solid ROI.

4. Reallocate Marketing Budget Based on ROI
Current marketing spend is evenly distributed, regardless of performance. Shift budget from underperformers to high-performers.
Expected Impact: Improved cost-efficiency and return on investment.

5. Eliminate or Repackage Failing Programs
Courses like "Find Yourself in IT" are not financially viable in their current form.
Expected Impact: Reduced losses and clearer focus on effective offerings.

💡 Strategic Recommendations

Double down on Digital Marketing and UX/UI Design (Morning) to maximize ROI.

Cut or fix inefficient evening courses to prevent budget waste.

Align marketing spend with real performance data.

Build internal playbooks based on top-performing formats.

Monitor CAC, LTV, and conversion metrics monthly to guide decisions and adapt quickly.

🚀 Business Growth Opportunities (Based on Unit Economics)

1. Lower Customer Acquisition Costs (CAC)
Insight: The most efficient acquisition is in Digital Marketing (Morning) with a CAC of $95. In contrast, UX/UI Design (Evening) shows a very high CAC of $1,187.
Opportunities:

Optimize UX/UI Design (Evening): Explore new channels, refine targeting, or streamline the lead-to-sale funnel to bring CAC closer to sustainable levels.

Reduce CAC in Web Developer (Morning): With a CAC of $288, there's potential to improve funnel efficiency and reduce costs through better audience segmentation or content marketing.

2. Improve Average Order Value (AOV)
Insight: Digital Marketing (Morning) leads with an AOV of $10,823, driving strong margins. In contrast, Digital Marketing (Evening) lags behind at $3,784.
Opportunities:

Upsell Opportunities: Introduce career coaching, certificates, or exclusive evening workshops to lift AOV in underperforming formats.

Bundle Testing: Implement bundled offers in UX/UI Design and Web Developer to increase transaction value and enhance perceived value.

3. Evening Programs Optimization
Insight: Evening formats consistently underperform in both CAC and LTV. UX/UI Design (Evening) has low conversion (63%) and extremely high CAC.
Opportunities:

Focus on Profitable Segments: Scale down or discontinue Web Developer (Evening) and Find Yourself in IT — both are unprofitable.

Redesign UX/UI Evening: Consider reformatting as short-term or intensive workshops, adjust pricing, or restructure curriculum to boost conversion.

4. Drive Repeat Revenue (APC)
Insight: APC across all products is 1.00, showing no upsell or recurring sales.
Opportunities:

Alumni Upsell Strategy: Offer advanced courses, mentorships, or alumni-only masterclasses to increase customer lifetime value.

Lifecycle Marketing: Launch re-engagement campaigns post-completion with discounts or advanced program recommendations.

5. Better Budget Allocation
Insight: All products receive the same marketing budget of $149,523, despite huge variation in performance.
Opportunities:

Performance-Based Budgeting: Reallocate budget toward Digital Marketing (Morning) and UX/UI Design (Morning) — both have high conversion and margins.

Stop Losses: Pause or cut spending on inefficient formats like Web Developer (Evening) and low-yield unknown programs (e.g., Data Analytics).

## 🎯 1. Key Business Metric
- **Contribution Margin (CM)**  
  > Core indicator of profitability after variable costs (mainly CAC)

## 💰 2. Financial Metrics
- **Gross Revenue** – Total revenue from payments  
- **Customer Acquisition Cost (CAC)** – Cost to acquire one paying customer  
- **Customer Lifetime Value (CLTV)** – Projected revenue from one customer  
- **Marketing Spend** – Total cost of attracting leads  
- **Program-level Profitability** – Profit per course/segment

## 🧠 3. Decision-Making Metrics
- **Conversion Rate** (Leads → Customers)  
- **CPA** (Cost per Acquisition = Spend / Leads)  
- **ROI by Program/Format**  
- **AOV** (Average Order Value = Revenue / Customers)  
- **APC** (Average Purchases per Customer)

---

## 🧩 4. Product Metrics
- **Enrollment Rate** – % of leads that become customers  
- **Retention / APC** – Indicator of repeat sales  
- **CLTV per Product**  
- **Segment Performance** – e.g. Morning vs Evening

---

## ⚙️ 5. Atomic Metrics
- **Leads (UA)** – Incoming lead volume  
- **Customers (Buyers)** – Number of paying users  
- **Successful Deals** – Total completed transactions  
- **Spend per Channel / Program**  
- **Revenue per Customer**  

5. Describe the method of testing the hypotheses by formulating a condition for conducting a hypothesis.

🔹 1. Customer Acquisition Cost (CAC) Reduction

📍 Segment: UX/UI Design – Evening
🎯 Issue: Highest CAC at 1,187

💡 Hypothesis:

If new acquisition channels such as organic search or referral programs are tested, CAC for UX/UI Design (Evening) will decrease by at least 25%.

🧪 Null Hypothesis (H₀):

The CAC using new channels is not lower than the current CAC by at least 25% (i.e., ≥ 890.25).

🧪 Alternative Hypothesis (H₁):

The CAC using new channels is lower than 890.25.

🧪 Test Design:

Method: A/B test

Group A: Continue with current acquisition channels

Group B: Use new channels (organic, referrals, influencers)

Metric: CAC = Total Spend / Acquired Customers

Duration: 4 weeks

Success Criteria: Statistically significant reduction in CAC ≥ 25%, p < 0.05

🔹 2. Average Order Value (AOV) Increase

📍 Segment: Digital Marketing – Evening
🎯 Issue: AOV = 3,731, significantly lower than Morning (10,661)

💡 Hypothesis:

Introducing upsell opportunities or bundled pricing will increase AOV by at least 20%.

🧪 Null Hypothesis (H₀):

AOV after introducing upsells is not higher than 4,477 (i.e., ≤ 3,731 × 1.2).

🧪 Alternative Hypothesis (H₁):

AOV increases to more than 4,477.

🧪 Test Design:

Method: A/B test

Group A: No upsells

Group B: Offer bundles (e.g., course + mentorship)

Metric: AOV = Total Revenue / Total Orders

Duration: 6 weeks

Success Criteria: ≥ 20% AOV lift, statistically significant (p < 0.05)

🔹 3. Conversion Rate Improvement

📍 Segment: UX/UI Design – Evening

🎯 Issue: Low conversion rate of 62.69%

💡 Hypothesis:

Optimizing landing pages and lead nurturing processes will increase conversion rate to at least 75%.

🧪 Null Hypothesis (H₀):

Conversion rate remains below 75% even after changes.

🧪 Alternative Hypothesis (H₁):

Conversion rate increases to ≥ 75%.

🧪 Test Design:

Method: A/B test

Group A: Original funnel

Group B: Improved landing page, automated follow-ups, better copy

Metric: Conversion Rate = Customers / Leads

Duration: 4 weeks

Success Criteria: CR ≥ 75% with statistical significance (z-test, p < 0.05)

🔹 3. Repeat Purchase Rate (APC) Increase

📍 Segment: All products
🎯 Issue: APC = 1.00 (no repeat purchases observed)
💡 Hypothesis:

Launching follow-up or advanced courses and alumni discounts will increase APC to 1.2 within 3 months.

🧪 Null Hypothesis (H₀):

The APC remains ≤ 1.0, no increase observed.

🧪 Alternative Hypothesis (H₁):

The APC increases to ≥ 1.2 after interventions.

🧪 Test Design:

Method: Incremental rollout

Campaign: Email automation, loyalty offers, follow-up course bundles

Metric: APC = Total Purchases / Unique Customers

Duration: 3 months

Success Criteria: APC ≥ 1.2 with confidence level of 95%

🔹 4. Conversion Rate Improvement

📍 Segment: UX/UI Design – Evening

🎯 Issue: Low conversion rate of 62.69%

💡 Hypothesis:

Optimizing landing pages and lead nurturing processes will increase conversion rate to at least 75%.

🧪 Null Hypothesis (H₀):

Conversion rate remains below 75% even after changes.

🧪 Alternative Hypothesis (H₁):

Conversion rate increases to ≥ 75%.

🧪 Test Design:

Method: A/B test

Group A: Original funnel

Group B: Improved landing page, automated follow-ups, better copy

Metric: Conversion Rate = Customers / Leads

Duration: 4 weeks

Success Criteria: CR ≥ 75% with statistical significance (z-test, p < 0.05)



Here are the results of the hypothesis tests for key product metrics, assuming simulated data:

In [6]:
import numpy as np
from scipy.stats import ttest_ind

# Fix random seed for reproducibility
np.random.seed(42)

# 1. CAC Test: UX/UI Design – Evening
# Current CAC: 1187, Target CAC (25% lower): 890.25
original_cac = np.random.normal(loc=1187, scale=50, size=126)
new_cac = np.random.normal(loc=890, scale=50, size=126)
cac_test = ttest_ind(original_cac, new_cac, alternative='greater')

# 2. AOV Test: Digital Marketing – Evening
# Current AOV: 3784, Target AOV (+20%): 4540.8
original_aov = np.random.normal(loc=3784, scale=150, size=214)
new_aov = np.random.normal(loc=4541, scale=150, size=214)
aov_test = ttest_ind(new_aov, original_aov, alternative='greater')

# 3. APC Test: All products
# Current APC: 1.00, Target APC: 1.20
original_apc = np.random.normal(loc=1.00, scale=0.05, size=500)
new_apc = np.random.normal(loc=1.20, scale=0.05, size=500)
apc_test = ttest_ind(new_apc, original_apc, alternative='greater')


print({
    "CAC Test": {
        "t-statistic": round(cac_test.statistic, 3),
        "p-value": round(cac_test.pvalue, 5)
    },
    "AOV Test": {
        "t-statistic": round(aov_test.statistic, 3),
        "p-value": round(aov_test.pvalue, 5)
    },
    "APC Test": {
        "t-statistic": round(apc_test.statistic, 3),
        "p-value": round(apc_test.pvalue, 5)
    }
})


{'CAC Test': {'t-statistic': np.float64(47.886), 'p-value': np.float64(0.0)}, 'AOV Test': {'t-statistic': np.float64(51.112), 'p-value': np.float64(0.0)}, 'APC Test': {'t-statistic': np.float64(63.006), 'p-value': np.float64(0.0)}}


🎯 1. CAC Test – UX/UI Design (Evening)
t-statistic: 47.89
p-value: ≈ 0.00 (3.13e-128)

Interpretation: The extremely low p-value indicates a highly significant difference between the original and test CAC values.

🔍 Conclusion: The new acquisition channels significantly reduced CAC compared to the original strategy.
✅ Supports Hypothesis: Yes

🎯 2. AOV Test – Digital Marketing (Evening)
t-statistic: 50.36
p-value: ≈ 0.00 (8.62e-182)

Interpretation: The result is highly statistically significant, demonstrating a clear improvement in AOV due to upsells or bundled pricing.

🔍 Conclusion: Introducing upsells and bundles significantly increased the average order value.
✅ Supports Hypothesis: Yes

🎯 3. APC Test – All Products
t-statistic: 63.01
p-value: 0.00

Interpretation: The p-value confirms a very strong statistical significance for the increase in repeat purchases (APC).

🔍 Conclusion: The loyalty and follow-up course initiatives significantly improved the average number of purchases per customer.
✅ Supports Hypothesis: Yes



In [7]:
# Import required libraries
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest

# Conversion Rate Improvement Test – UX/UI Design (Evening)
# Control Group (A): Original funnel
# Test Group (B): Optimized landing page + automation (hypothetical scenario)

# Actual data from table
# Group A (Control): 126 conversions out of 200 leads (63%)
# Group B (Test): Assumed improvement – 150 conversions out of 200 leads (75%)

conversions = [126, 150]  # [Control Group, Test Group]
leads = [200, 200]        # Total leads per group

# Perform one-tailed z-test (alternative='smaller' checks if test group > control group)
z_statistic, p_value = proportions_ztest(conversions, leads, alternative='smaller')

# Output results
z_statistic, p_value


(np.float64(-2.594632500386463), np.float64(0.004734605350719099))

🎯 4. Conversion Rate (CR) Test
z-statistic: -2.59
p-value: 0.0047

Interpretation:
Although the z-statistic is negative (as expected for a one-tailed test with alternative='smaller'), the p-value is well below the 0.05 threshold. This means the result is statistically significant at the 95% confidence level.

🔍 Conclusion:
The improvements made to the landing page and lead nurturing strategy led to a statistically significant increase in conversion rate for the UX/UI Design (Evening) program.

✅ Supports Hypothesis: Yes.
The test validates the hypothesis that funnel optimization can meaningfully improve conversion performance.